functions/Export-DbaDacpac.ps1

```diff
@@ -0,0 +1,179 @@
+	)
+	
+	process {
+		if (!$Database -and !$ExcludeDatabase -and !$AllUserDatabases) {
+			Stop-Function -Message "You must specify databases to execute against using either -Database, -ExcludeDatabase or -AllUserDatabases" -Continue
+		}
+		
+		if (-not (Test-Path $Directory)) {
+			Stop-Function -Message "$Directory doesn't exist or access denied" -Continue
+		}
+		
+		foreach ($instance in $sqlinstance) {
+			
+			try {
+				Write-Message -Level Verbose -Message "Connecting to $instance."
+				$server = Connect-SqlInstance -SqlInstance $instance -SqlCredential $sqlcredential -MinimumVersion 9
+			}
+			catch {
+				Stop-Function -Message "Failure" -Category ConnectionError -ErrorRecord $_ -Target $instance -Continue
+			}
+			
+			$dbs = $server.Databases | Where-Object { $_.IsSystemObject -eq $false }
+			
+			if ($Database) {
+				$dbs = $dbs | Where-Object Name -in $Database
+			}
+			
+			if ($ExcludeDatabase) {
+				$dbs = $dbs | Where-Object Name -notin $ExcludeDatabase
+			}
+			
+			foreach ($db in $dbs) {
+				$dbname = $db.name
+				$connstring = $server.ConnectionContext.ConnectionString.Replace('"', "'")
+				if ($connstring -notmatch 'Database=') {
+					$connstring = "$connstring;Database=$dbname"
+				}
+				$path = "$Directory\$dbname.dacpac"
+				Write-Message -Level Verbose -Message "Exporting $path"
+				Write-Message -Level Verbose -Message "Using connection string $connstring"
+				
+				$sqlPackageArgs = "/action:Extract /tf:""$path"" /SourceConnectionString:""$connstring"" $ExtendedParameters $ExtendedProperties"
+				
+				try {
+					$resultstime = [diagnostics.stopwatch]::StartNew()
+					
+					$startprocess = New-Object System.Diagnostics.ProcessStartInfo
+					$startprocess.FileName = "$script:PSModuleRoot\bin\dacfx\sqlpackage.exe"
+					$startprocess.Arguments = $sqlPackageArgs
+					$startprocess.RedirectStandardError = $true
+					$startprocess.RedirectStandardOutput = $true
+					$startprocess.UseShellExecute = $false
+					$process = New-Object System.Diagnostics.Process
+					$process.StartInfo = $startprocess
+					$process.Start() | Out-Null
+					$process.WaitForExit()
+					$stdout = $process.StandardOutput.ReadToEnd()
+					$stderr = $process.StandardError.ReadToEnd()
+					Write-Message -level Verbose -Message "StandardOutput: $stdout"
+					
+					[pscustomobject]@{
+						ComputerName	 = $server.NetName
+						InstanceName	 = $server.ServiceName
+						SqlInstance	     = $server.DomainInstanceName
+						Database		 = $dbname
+						FileName	   = $path
+						Elapsed	       = [prettytimespan]($resultstime.Elapsed)
+						Status		   = "Success"
+					} | Select-DefaultView -ExcludeProperty ComputerName, InstanceName
+				}
+				catch {
+					[pscustomobject]@{
+						ComputerName	 = $server.NetName
+						InstanceName	 = $server.ServiceName
+						SqlInstance	     = $server.DomainInstanceName
+						Database		 = $dbname
+						FileName	   = $path
+						Elapsed	       = [prettytimespan]($elapsed)
```

Comment:
I don't see `$elapsed` defined.
Could the `$resultstime = [diagnostics.stopwatch]::StartNew()` be moved outside the `try{} catch{}` so we have a value for 'elapsed' whether it fails or succeeds?

Reply:
I was surprised (and disappointed) to find out that it does keep the variable, no matter what. at least that was the behavior last time I tested it. but I'll move it above try.